src/viur/core/modules/history.py

```diff
@@ -0,0 +1,596 @@
+        assert self.client and self.table
+        if res := self.client.insert_rows(self.table, [data]):
+            raise ValueError(res)
+
+
+class HistoryAdapter(DatabaseAdapter):
+    """
+    Generalized adapter for handling history events.
+    """
+
+    DEFAULT_EXCLUDES = {
+        "key",
+        "changedate",
+        "creationdate",
+        "importdate",
+        "viurCurrentSeoKeys",
+    }
+    """
+    Bones being ignored within history.
+    """
+
+    def __init__(self, excludes: t.Iterable[str] = DEFAULT_EXCLUDES):
+        super().__init__()
+
+        # add excludes to diff excludes
+        self.diff_excludes = set(excludes)
+
+    def prewrite(self, skel, is_add, change_list=()):
+        if not is_add:  # edit
+            old_skel = skel.clone()
+            old_skel.read(skel["key"])
+            self.trigger("edit", old_skel, skel, change_list)
+
+    def write(self, skel, is_add, change_list=()):
+        if is_add:  # add
+            self.trigger("add", None, skel)
+
+    def delete(self, skel):
+        self.trigger("delete", skel, None)
+
+    def trigger(
+            self,
+            action: str,
+            old_skel: SkeletonInstance,
+            new_skel: SkeletonInstance,
+            change_list: t.Iterable[str] = (),
+    ) -> str | None:
+
+        # skip excluded actions like login or logout
+        if action in conf.history.excluded_actions:
+            return None
+
+        # skip when no user is available or provided
+        if not (user := current.user.get()):
+            return None
+
+        # FIXME: Turn change_list into set, in entire Core...
+        if change_list and not (change_list := set(change_list).difference(self.diff_excludes)):
+            logging.info("change_list is empty, nothing to write")
+            return None
+
+        # skip excluded kinds and history kind to avoid recursion
+        any_skel = (old_skel or new_skel)
+        if any_skel and (kindname := getattr(any_skel, "kindName", None)):
+            if kindname in conf.history.excluded_kinds:
+                return None
+
+            if kindname == "viur-history":
+                return None
+
+        return conf.main_app._viur_history.write_diff(
+            action, old_skel, new_skel,
+            change_list=change_list,
+            user=user,
+            diff_excludes=self.diff_excludes,
+        )
+
+
+class ViurHistory(List):
+    """
+    ViUR history module
+    """
+    kindName = "viur-history"
+
+    adminInfo = {
+        "name": "History",
+        "icon": "clock-history",
+        "filter": {
+            "orderby": "timestamp",
+            "orderdir": "1",
+        },
+        "disabledActions": ["add", "clone", "delete"],
+    }
+
+    roles = {
+        "admin": "view",
+    }
+
+    HISTORY_VERSION = 1
+    """
+    History format version.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.bigquery = BigQueryHistory()
```

Comment:
This should be customizable and sub-classable via a `History.BigQueryHistoryCls` member.
```py
   BigQueryHistoryCls = BigQueryHistory
```
and then
```suggestion
        self.bigquery = self.BigQueryHistoryCls and self.BigQueryHistoryCls()
```
which will also allow to entirely unset the BigQueryHistory-adapter.